src/util/jsonify.ts

```diff
@@ -0,0 +1,76 @@
+import {
+	Decimal,
+	Duration,
+	Geometry,
+	RecordId,
+	StringRecordId,
+	Table,
+	Uuid,
+} from "../data";
+
+export type Jsonify<T> = T extends
+	| Date
+	| Uuid
+	| Decimal
+	| Duration
+	| StringRecordId
+	? string
+	: T extends undefined
+		? never
+		: T extends Record<string | number | symbol, unknown> | Array<unknown>
+			? { [K in keyof T]: Jsonify<T[K]> }
+			: T extends Map<infer K, infer V>
+				? Map<K, Jsonify<V>>
+				: T extends Set<infer V>
+					? Set<Jsonify<V>>
+					: T extends Geometry
+						? ReturnType<T["toJSON"]>
+						: T extends RecordId<infer Tb>
+							? `${Tb}:${string}`
+							: T extends Table<infer Tb>
+								? `${Tb}`
+								: T;
+
+export function jsonify<T>(input: T): Jsonify<T> {
+	if (typeof input === "object") {
+		if (input === null) return null as Jsonify<T>;
+
+		// We only want to process "SurrealQL values"
+		if (
+			input instanceof Date ||
+			input instanceof Uuid ||
+			input instanceof Decimal ||
+			input instanceof Duration ||
+			input instanceof StringRecordId ||
+			input instanceof RecordId ||
+			input instanceof Geometry ||
+			input instanceof Table
+		) {
+			return input.toJSON() as Jsonify<T>;
+		}
+
+		// We check by prototype, because we do not want to process derivatives of objects and arrays
+		switch (Object.getPrototypeOf(input)) {
```

Comment:
What about null prototype?

Reply:
The first check in the function checks for null, so that's already filtered out at this point